contrib/babelfishpg_tsql/src/pltsql_coerce.c

```diff
@@ -888,6 +921,227 @@ run_tsql_best_match_heuristics(int nargs, Oid *input_typeids, FuncCandidateList
 	return new_candidates;
 }
 
+/*
+ * get_immediate_base_type_of_UDT_internal()
+ * This function returns the Immediate base type for UDT.
+ * Returns InvalidOid if given type is not an UDT
+ */
+static Oid
+get_immediate_base_type_of_UDT_internal(Oid typeid)
```

Comment:
Consider following example, 
```
— When immediate base type is used for validation
1> CREATE TYPE dbo.babel_4489_trim_imageUDT FROM image;
2> go
1> create table t(a dbo.babel_4489_trim_imageUDT)
2> go
1> insert into t values(0x616263)
2> go

(1 rows affected)
1> select trim(a from a) from t
2> go
Msg 33557097, Level 16, State 1, Server BABELFISH, Line 1
Argument data type image is invalid for argument 1 of Trim function.


— When bottom base type is used for validation
1> CREATE TYPE dbo.babel_4489_trim_imageUDT FROM image;
2> go
1> create table t(a dbo.babel_4489_trim_imageUDT)
2> go
1> insert into t values(0x616263)
2> go

(1 rows affected)
1> select trim(a from a) from t
2> go
Msg 33557097, Level 16, State 1, Server BABELFISH, Line 1
Argument data type bbf_varbinary is invalid for argument 1 of Trim function.
```
Here in the above example, we can see that for UDT on image, error thrown is `Argument data type bbf_varbinary is invalid for argument 1 of Trim function.` i.e. varbinary is used for datatype validation, instead image should have been used. Hence there is a need for immediate base type and not the bottom base type.

Reply:
Not exactly, when using bottom base type if we consider different string function which support `varbinary` as an argument but does not accept `image` as argument. We will get unexpected output instead of error, as since the validation happened on bottom base type which will be `varbinary` and it will get through validation part but instead validation should have been done on `image` (immediate base type). Hence we need `bbf_ get_immediate_base_type_of_UDT_internal` function.